File: 0139-word-break/0139-word-break.java
class Solution {
    boolean done = false;
    HashSet<Integer> hs = new HashSet<Integer>();
    public boolean wordBreak(String s, List<String> wordDict) {
        dp(s, 0, 1, wordDict);
        return done;
    }
    
    public void dp(String s, int a, int b, List<String> wordDict)
    {
        if ((wordDict.contains(s.substring(a,b)) && b == s.length()) || done == true)
        {
            done = true;
            return;
        }
        if (b==s.length())
        {
            hs.add(a);
            return;
        }
        if (hs.contains(a))
        {
            return;
        }
        if (wordDict.contains(s.substring(a,b)))
        {
            dp(s,b,b+1,wordDict);
        }
        // System.out.println(s.substring(a,b))
        dp(s,a,b+1,wordDict);
        return;
    }
}

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        temp = [False] * (len(s) + 1)
        temp[len(s)] = True
        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if (i + len(w) <= len(s)) and s[i:i+len(w)] == w:
                    temp[i] = temp[i + len(w)]
                if temp[i] == True:
                    break
            print(temp)
        return temp[0]